src/formatting/generated.rs

```diff
@@ -0,0 +1,40 @@
+use std::{
+    fs,
+    io::{self, BufRead},
+};
+
+use crate::config::file_lines::FileName;
+use crate::formatting::comment::contains_comment;
+
+/// Returns `true` if the given span is a part of generated files.
+pub(super) fn is_generated_file(file_name: &FileName, original_snippet: Option<&String>) -> bool {
+    let first_line = match file_name {
+        FileName::Stdin => original_snippet
+            .and_then(|s| s.lines().next())
+            .map(str::to_owned)
+            .unwrap_or("".to_owned()),
+        FileName::Real(ref path) => fs::File::open(path)
```

Comment:
Not ideal to have another file system read (IIRC we're up to at least 3 due with this plus the parser and one for newline checks) :thinking:  

Would it be feasible to get the first line without the file system hit, maybe via the parse session data? Pperhaps it won't matter that much given the default is false, and/or could potentially be updated later

Reply:
(forgive me for the intrusion, i was recently doing work in this area and learned this: unless you're very low on memory, recently-accessed files are cached by the kernel and effectively free to access; the slowest part will be parsing)

Reply:
> forgive me for the intrusion

no worries at all!

> the slowest part will be parsing

agreed, that's basically a given, but my question wasn't really about a parsing/AST construction vs. additional file reads, it's whether the additional file read is necessary given that the file's already been read and loaded from the preceding parsing stage. 

> unless you're very low on memory, recently-accessed files are cached by the kernel and effectively free to access;

do you know if that holds true everywhere (for example Windows, some of the tier 2/3 targets, etc.)?

I'm not terribly familiar in that space, so entirely possible this approach is better than an alternative like using the span from the associated mod to get the first line from the snippet for the file, either more performant due to the caching or other reasons, and/or simpler from a code readability and management perspective :man_shrugging: 

Also readily admit such a question may be unnecessary/premature optimization, but figured I'd ask anyway :smile:

Reply:
yeah, I agree that reusing data that is already and easily available is preferable. I just wanted to note that the performance implications of rereading a file may not be as high as it may appear, and that there are probably bigger bottlenecks. But I totally agree with you, no need to do extra work if you don't have to.

Re environments: most modern kernels (~last 20 years) have a page cache; this applies to windows as well

Reply:
> Re environments: most modern kernels (~last 20 years) have a page cache; this applies to windows as well

Sure, but I suppose I should've phrased my question better.

Was wondering more so if you thought there'd always be sufficient available memory to guarantee/near-guaranatee cache hit in all scenarios. I know even on my laptops I can get really memory constrained (as I tend to use lower end machines and still go overboard with multi tasking :laughing:), but a lot of CI environments can be pretty low on resources and one of the things I learned during the conversations on the compiler frontend for rls/ra is that there's plenty of folks doing their inner dev loop on some pretty resource constrainted environments, and even some remote workflows.

I genuinely don't know, but wouldn't be surprised if running in some of the more resource constrained environments (or those that purged more aggressively) could result in some cache misses and i/o hits. 

I'm quickly getting out my depth though and will happily defer to @topecongiro on whether it's worth considering an alternative :smile:

Reply:
I first tried by using the span of `ast::Mod`, though it turned out that the span may not contain the comment at the beginning or the end of the file. I think that there are ways to get a correct span, though I opted for a simpler implementation this time.

Reply:
Works for me!